sheerwater_benchmarking/utils/caching.py

```diff
@@ -305,28 +306,90 @@ def read_from_delta(cache_path):
+
+        if not isinstance(df, dd.DataFrame):
+            raise RuntimeError("Upsert is only supported by dask dataframes for parquet")
+
+        # Check to see if the cache exists
+        if fs.exists(verify_path) and fs.exists(cache_path):
+            if fs.exists(verify_path):
+                fs.rm(verify_path, recursive=True)
+
+            print("Found existing cache for upsert.")
+            existing_df = read_from_parquet(cache_path)
+
+            # remove any rows already in the dataframe
+            outer_join = existing_df.merge(df, how = 'outer', indicator = True)
+            new_rows = outer_join[(outer_join._merge == 'right_only')].drop('_merge', axis = 1)
+
+            # Coearce dtypes - may not be necessary?
+            # new_rows = new_rows.astype(existing_df.dtypes)
+
+            # write in append mode
+            print("Appending new rows to existing parquet.")
+            new_rows.to_parquet(
+                cache_path,
+                overwrite=False,
+                append=True,
+                partition_on=part,
+                engine="pyarrow",
+                write_metadata_file=True,
+                write_index=False,
+            )
+
+            fs.open(verify_path, 'w').write(datetime.datetime.now(datetime.timezone.utc).isoformat())
+        else:
+            if fs.exists(verify_path):
+                fs.rm(verify_path, recursive=True)
+
+            # If it doesn't just write
+            print(f"Cache {cache_path} doesn't exist for upsert.")
+            df.to_parquet(
+                cache_path,
+                overwrite=overwrite,
+                partition_on=part,
+                engine="pyarrow",
+                write_metadata_file=True,
+                write_index=False,
+            )
+
+            fs.open(verify_path, 'w').write(datetime.datetime.now(datetime.timezone.utc).isoformat())
     else:
-        raise ValueError("Can only write dask and pandas dataframes to parquet.")
+        if fs.exists(verify_path):
+            fs.rm(verify_path, recursive=True)
+
+        if fs.exists(cache_path):
+            fs.rm(cache_path, recursive=True)
+
+        if isinstance(df, dd.DataFrame):
+            df.to_parquet(
+                cache_path,
+                overwrite=overwrite,
+                partition_on=part,
+                engine="pyarrow",
+                write_metadata_file=True,
+                write_index=False,
+            )
+        elif isinstance(df, pd.DataFrame):
+            df.to_parquet(cache_path, partition_cols=part, engine='pyarrow')
+        else:
+            raise ValueError("Can only write dask and pandas dataframes to parquet.")
 
-    # Add a lock file to the cache to verify cache integrity, with the current timestamp
-    fs.open(verify_path, 'w').write(datetime.datetime.now(datetime.timezone.utc).isoformat())
+        fs.open(verify_path, 'w').write(datetime.datetime.now(datetime.timezone.utc).isoformat())
```

Comment:
Could move this verify logic out of the if statements and just have it once at the bottom